### ISSUE TITLE
Error in src/rdb_protocol/datum.cc at line 1470: Unreachable code

### ISSUE THREAD

**Issue by tatsujin1** [rethinkdb/rethinkdb#5438]
v2.2.3 on debian 8.2

After it was automatically restarted (by systemd) it seems to work fine.

If the crash was caused by some specific data in the db I unfortunately have no way to find what that could've been.

```
2016-02-23T12:38:00.750202800 3470395.004338s error: Error in src/rdb_protocol/datum.cc at line 1470:
2016-02-23T12:38:00.750480828 3470395.004616s error: Unreachable code:
2016-02-23T12:38:00.750521538 3470395.004656s error: Backtrace:
2016-02-23T12:38:01.079540958 3470395.333680s error: Tue Feb 23 12:38:00 2016

1 [0xa8b5f0]: backtrace_t::backtrace_t() at ??:?
2 [0xa8b969]: format_backtrace(bool) at ??:?
3 [0xd00a9d]: report_fatal_error(char const*, int, char const*, ...) at ??:?
4 [0x8e26be]: void ql::datum_t::write_json_unchecked_stack<rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator> >(rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator>*) const at ??:?
5 [0x8e3845]: int call_with_enough_stack<int, void call_with_enough_stack<void ql::datum_t::write_json<rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator> >(rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator>*) const::{lambda()#1}>(void ql::datum_t::write_json<rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator> >(rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator>*) const::{lambda()#1}&&, unsigned long)::{lambda()#1}>(void call_with_enough_stack<void ql::datum_t::write_json<rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator> >(rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator>*) const::{lambda()#1}>(void ql::datum_t::write_json<rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator> >(rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator>*) const::{lambda()#1}&&, unsigned long)::{lambda()#1}&&, unsigned long) at ??:?
6 [0x8e38d1]: void ql::datum_t::write_json<rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator> >(rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::RAllocator>*) const at ??:?
7 [0xa806b8]: write_response_internal(ql::response_t*, rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::RAllocator>*, bool) at ??:?
8 [0xa8253d]: json_protocol_t::send_response(ql::response_t*, long, linux_tcp_conn_t*, signal_t*) at ??:?
9 [0xa882d9]: void save_exception<void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda()#1}::operator()() const::{lambda()#1}>(std::__exception_ptr::exception_ptr*, std::string*, cond_t*, void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda()#1}::operator()() const::{lambda()#1}&&) at ??:?
10 [0xa8884a]: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda()#1}::operator()() const at ??:?
11 [0x99c518]: coro_t::run() at ??:?
2016-02-23T12:38:01.079807642 3470395.333942s error: Exiting.
```

edit: grammar

**Comment by danielmewes** [comment]
Thanks for the report @tatsujin1 . I've not seen this crash before.
Did it leave behind a core file by any chance?

**Comment by tatsujin1** [comment]
Oh, didn't think to check for that. I'll look for it tomorrow.

**Comment by tatsujin1** [comment]
Unfortunately no core file was to be found :-(

The syslog also contained this (apart from the above stacktrace):
```
Feb 23 12:38:01 xxxx systemd[1]: rethinkdb@xxxx.service: Main process exited, code=killed, status=5/TRAP
Feb 23 12:38:01 xxxx systemd[1]: rethinkdb@xxxx.service: Unit entered failed state.
Feb 23 12:38:01 xxxx systemd[1]: rethinkdb@xxxx.service: Failed with result 'signal'.
```
Not very helpful either, I guess...

Is there anything I can do to make the next crash (if any) more informative?

**Comment by tatsujin1** [comment]
Just noticed that it happened again this morning.

**Comment by danielmewes** [comment]
It might be helpful to enable core files. You can usually do this by editing /etc/security/limits.conf, and adding a line like this:
```
*               soft    core            100000000
```

(though I haven't tried this particular configuration, so I'm not sure whether it gets applied correctly to rethinkdb when started through a init script or not)


The fact that the same crash happened again might hint at faulty data in one of your tables.
I suggest running the following query to find out which table (if any) is crashing the server:
```js
r.db(...).table(...).map(r.row.toJsonString()).count()
```
You can run this from the Data Explorer. It might time out if the tables are large, but the query will actually keep running in the background (watch the reads/s graph).
*Warning:* If there is indeed a faulty document in a table, running this query will cause the server to crash again. I recommend only running this at a time where an outage is acceptable.

Once we know if the faulty data is in a table and in which table it is, is there a chance that you could send us a copy of that table? I can give you more detailed instructions and a place that you can upload it to later. We can also sign an NDA for the table's data if necessary.

A few general questions:
What hardware is this running on (any virtualization?)?
What file system are you using for the RethinkDB data partition? Are there any special mount options?
Is this a single server or a cluster? If it's a cluster, how many nodes?

**Comment by tatsujin1** [comment]
Ah, so that's where that core setting is nowadays! `ulimit` doesn't seem to exist anymore(?)... Well, anyway, I'll add that tomorrow.

Oh I see, that's probably a simple way to find which table the problem is in, if there is any. But the result of that test is basically boolean, right? I'll probably do a variant of that, that returns one document at a time to the client, using a for-loop over all documents. Luckily this database (and clients) isn't completely critical (yet), and all related services will be automatically restarted by systemd when the DB crashes. And it's all set up to synchronize their state on startup, so there's likely not too much data-loss :) But I'll see if I can get around to doing something like this tomorrow.

Could something like an invalid utf-8 sequence cause this? Non-ascii text is quite common, and the content is from external sources (M2M). It might in theory contain all kinds of "things"... However, the data is received in json format which is parsed by rapidjson (in python), so I hope it's safe to assume it is valid.

But, reading the stack trace it seems that it crashed "on the way up" from a write; generating the response to the client. Does this imply that the data was written to the DB? Hm... but wait, we use soft durability so I guess it was never persisted to disk in any case... And also the changefeeds likely never got any message for that document before the db crashed.

The configuration is quite simple; a single node on a 4-core VM. 8G ram, I think.
File system and most other things is basically untouched debian defaults. I even think the database is on the root partition (!).
The db load is about 10 updates per second and then probably about 20 changefeeds on that with various queries. The biggest table - where most of the action takes place - is probably between 50-100k documents of ~4kB.

**Comment by danielmewes** [comment]
> Oh I see, that's probably a simple way to find which table the problem is in, if there is any.

Right. I was hoping that once we know the table, you could maybe provide the data to us so we can analyze the issue further on our end.

> Could something like an invalid utf-8 sequence cause this?

It should not. In general there's nothing you can send to the server that *should* cause this, so it's either a server bug (likely) or some form of file system and/or memory corruption (less likely, though it depends a bit on the storage virtualization layer). Your guess as to what might have triggered this is as good as mine.

It's plausible that there's an issue with our JSON parser and/or validation that allows illegal data to get through the query layer.


The data size and load that you describe sound comparably light. Which virtualization solution are you using?

**Comment by danielmewes** [comment]
Note that the backtrace doesn't necessarily indicate that the query triggering the crash was a write. `write_response_internal` is used for writing the protocol response for all types of queries, both reads and writes, and even changefeeds.

**Comment by tatsujin1** [comment]
Ah yes, true.. Hm, so it could be, as you say, anything...
The virtualization is vmware.

**Comment by tatsujin1** [comment]
> Right. I was hoping that once we know the table, you could maybe provide the data to us so we can analyze the issue further on our end.

Ok. That's probably not a problem.

**Comment by tatsujin1** [comment]
I've now tried the map-to-json thingy, but nothing happened. Actually, the query was quite quick, 60k documents in ~1 second. Does the `count()` optimize away the `map()` somehow?

Anyway, I've applied the core size limit to the running processes, so I hope we'll get a core next time around. It does seem to happen at least daily.

**Comment by tatsujin1** [comment]
Even though I set the core size limit (using `prlimit`) it won't create a core file :(

**Comment by tatsujin1** [comment]
One detail we've noticed from the last two days of crashes is that it is mysteriously repetitive.
```
2016-02-24T09:41:28.542232474 5.012031s notice: Running rethinkdb 2.2.3+1~0jessie (GCC 4.9.2)...
2016-02-24T14:21:52.637779099 5.012659s notice: Running rethinkdb 2.2.3+1~0jessie (GCC 4.9.2)...
2016-02-25T09:43:06.950116060 5.012061s notice: Running rethinkdb 2.2.3+1~0jessie (GCC 4.9.2)...
2016-02-25T14:30:55.886399318 5.014001s notice: Running rethinkdb 2.2.3+1~0jessie (GCC 4.9.2)...
```
We have no explanation of why this might be, yet.
Despite having plentiful of graphs over e.g. network & cpu loads, we can see no correlation, yet.

**Comment by danielmewes** [comment]
> I've now tried the map-to-json thingy, but nothing happened. Ran through all tables without a hitch.

That sounds like the stored data is ok then.

So it's probably some query that triggers this. I wish we had a query log to debug things like this, but unfortunately we don't have that yet. :-(

I realize that this might be a lot of work, so we can think of some other options for debugging this if it's too difficult: Is it easily possible in your application to add some form of query log on the application side? I.e. whenever you call `run`, write the query itself into a log file together with the current date and time? (you should be able to convert the query object that you call `run` on to a string)
That could give us some hints as to what's triggering the crash.

**Comment by tatsujin1** [comment]
Good idea. Turns out it was quite simple (after all, it's python! :)... I simply monkey-patched the `_start()` method on the `Connection` class to do the logging.
But man, that is a _lot_ of queries!  :)

Let's see if it crashes at ~9:40 today also, maybe we'll get some useful information...

**Comment by tatsujin1** [comment]
Indeed it crashed as expected, promptly at 09:37... almost like clockwork.
I now have a bunch of logs from the various processes that did queries. All but one, which is a read-only nodejs server, which I - for now - assume is not responsible.

There are no obvious queries that stand out; nothing that also happen about every other second...
I think we'll wait for the 14:30 crash and compare the logs.

**Comment by tatsujin1** [comment]
Of course it crashed again ~14:30, you could almost set a watch by these crashes...
The query logs contains the same similar standard queries; mostly `get()`, `get_all()`, slightly fewer `update()`, `delete()` and the occasional `insert()`...

I also did a `top -b -d 1` before (and during) the crash; cpu load was regular ups & downs throughout. The 1-minute average system load climed during the last 3 minutes from 0.1 to 0.5. But then again, it is within normal fluctuations.

Our best correlation so far is that the times kind of coincide with coffee breaks. ;)

**Comment by danielmewes** [comment]
Very odd so far. Thanks for sharing all the info.

I don't have any great explanations yet. I'll take a look at some code later and see if I can come up with any path that could lead to this. Unfortunately the surface area of code that feeds data into the place where it crashes (according to the backtrace) is rather huge.

**Comment by tatsujin1** [comment]
It will be interesting to see if/how the times of the crashes will differ during the weekend.

**Comment by tatsujin1** [comment]
In recent discoveries has found that it is related to changefeed. This morning (ahead of the "morning crash"), we disabled a web view (nodejs, which uses changefeeds to push updates to clients). No crash!

**Comment by tatsujin1** [comment]
With version 2.2.5 it's very easy to reproduce using these scripts:
**test-modify-5438.py:**
```
#! /usr/bin/env python3

import rethinkdb as r
import random
import time

r.connect().repl()

db_name = 'test_db'
table_name = 'test_table'

if db_name not in r.db_list().run():
	r.db_create(db_name).run()

if table_name not in r.db(db_name).table_list().run():
	r.db(db_name).table_create(table_name).run()

table = r.db(db_name).table(table_name)
table.config().update({ 'durability': 'soft' }).run()

print('cleaning table...')
table.delete().run(noreply=True)

indexes = [	'index' ]
existing_indexes = table.index_list().run()
for index in [index for index in indexes if index not in existing_indexes]:
	table.index_create(idx).run()

max_id = 5000
print('inserting...', max_id)
for doc_id in range(1, max_id + 1):
	table.insert({
		'id': doc_id,
		'index': random.randint(1, 10),
		'value1': random.randint(1, 10),
	}).run(noreply=True)


while True:
	time.sleep(0.05)

	doc_id = random.randint(1, max_id)

	existing = table.get(doc_id).run()
	if existing:
		if random.random() > 0.2:
			table.get(doc_id).update({
				'index': random.randint(1, 10)
			}).run(noreply=True)
			print('   up   :', doc_id)

		else:
			table.get(doc_id).delete().run(noreply=True)
			print('     del:', doc_id)

	else:
		table.insert({
			'id': doc_id,
			'index': random.randint(1, 10),
			'value1': random.randint(1, 10),
		}).run(noreply=True)
		print('ins     :', doc_id)
```
Start that and leave running.

Then start the following using nodejs:
**test-5438.js:**
```

var r = require('rethinkdb');

var query = r.db('test_db').table('test_table')
	.getAll(1, 2, 3, 4, 5, { index: 'index' })
	.filter(r.row('value1').gt(2));

r.connect(function(err, conn) {
	var first = true;
	console.time('first message');
	query.changes().run(conn, function(err, cursor) {
		cursor.on('data', function(item) {
			if(first)
			{
				console.timeEnd('first message');
				first = false;
			}
			else
				process.stdout.write('.');
		});
		cursor.on('error', function(err) {
			console.log('ERROR: ' + err.message);
		});
	});
});
```
Almost the instant the nodejs script is started the rethinkdb server crashes. Both the `getAll()` and `filter()` parts of the query are required.

However, so far I've only seen the crash happen with v2.2.5.
Running v2.2.3 (the version we currently use in production) or v2.2.4 does not crash using this test.

**Comment by tatsujin1** [comment]
Success! I've now reproduced the crash with v2.2.3 as well.
Hopefully it's the same bug that crashes both versions (the stack trace is almost identical), but anyway here are the variants that crashes v2.2.3:
**test-modify-5438.py:**
```
#! /usr/bin/env python3

import rethinkdb as r
import random
import time

r.connect().repl()

db_name = 'test_db'
table_name = 'test_table'

if db_name not in r.db_list().run():
	r.db_create(db_name).run()

if table_name not in r.db(db_name).table_list().run():
	r.db(db_name).table_create(table_name).run()

table = r.db(db_name).table(table_name)
table.config().update({ 'durability': 'soft' }).run()

print('cleaning table...')
table.delete().run(noreply=True)

indexes = [	'index' ]
existing_indexes = table.index_list().run()
for index in [index for index in indexes if index not in existing_indexes]:
	table.index_create(idx).run()

max_id = 5000
print('inserting...', max_id)
for doc_id in range(1, max_id + 1):
	table.insert({
		'id': doc_id,
		'index': random.randint(1, 10),
		'value1': random.randint(1, 10),
		'value2': random.randint(1, 10),
		'value3': random.randint(1, 10),
	}).run(noreply=True)


while True:
	time.sleep(0.05)

	doc_id = random.randint(1, max_id)

	existing = table.get(doc_id).run()
	if existing:
		if random.random() > 0.2:
			table.get(doc_id).update({
				'index': random.randint(1, 10),
				'value1': random.randint(1, 10),
				'value2': random.randint(1, 10),
				'value3': random.randint(1, 10),
			}).run(noreply=True)
			print('   up   :', doc_id)

		else:
			table.get(doc_id).delete().run(noreply=True)
			print('     del:', doc_id)

	else:
		table.insert({
			'id': doc_id,
			'index': random.randint(1, 10),
			'value1': random.randint(1, 10),
			'value2': random.randint(1, 10),
			'value3': random.randint(1, 10),
		}).run(noreply=True)
		print('ins     :', doc_id)
```
It seems that it has to do with the `update()` part. Commenting out the value1/2/3 fields from the update averts the crash.

And the nodejs script:
**test-5438.js:**
```
'use strict';

var r = require('rethinkdb');

var query = r.db('test_db').table('test_table')
	.getAll(1, 2, 3, 4, 5, { index: 'index' })
	.filter(r.row('value1').gt(2))
	.filter(r.row('value2').lt(8))
	.filter(r.row('value3').gt(2))

r.connect(function(err, conn) {
	var first = true;
	console.time('first message');
	query.changes().run(conn, function(err, cursor) {
		cursor.on('data', function(item) {
			if(first)
			{
				console.timeEnd('first message');
				first = false;
			}
			else
				process.stdout.write('.');
		});
		cursor.on('error', function(err) {
			console.log('ERROR: ' + err.message);
		});
	});
});
```

**Comment by danielmewes** [comment]
Thanks so much for working on a test case and sharing it with us @tatsujin1 . That is *extremely* helpful.

The one thing we changed about changefeeds in 2.2.5 is https://github.com/rethinkdb/rethinkdb/issues/5205 , so I suspect that this is related. Though we'll still need to understand why it sometimes also happened with 2.2.3. Maybe the cause is the same, but the change in 2.2.5 makes it easier to hit?

@mlucy can you look into this please?

**Comment by tatsujin1** [comment]
Yes, seems so. The crash in 2.2.3 usually takes a few seconds after the nodejs script is started to crash, while 2.2.5 crashes almost instantly.

**Comment by mlucy** [comment]
Thanks for the great test cases @tatsujin1 !

I have a fix for this in branch `mlucy_5483`, CR 3508 by @danielmewes .

**Comment by tatsujin1** [comment]
Great news!   Any idea when this will be included in a release?

**Comment by danielmewes** [comment]
@tatsujin1 This will be in RethinkDB 2.2.6, but it might take us two weeks or so to get out. Since you're running into this frequently, I can send you a pre-release package based on 2.2.5 with the patch applied. Could you drop me a quick email to daniel@rethinkdb.com ?

@jodagm could you send me a short email as well please? You're running Ubuntu Wily right?

**Comment by tatsujin1** [comment]
Unfortunately, for some mysteriously and as-of-yet discovered reason, we haven't had any crash since sunday...
So I can't really say if this fix fixed the problem for us. :-|

**Comment by mlucy** [comment]
This fix is in v2.2.x and next.

**Comment by gngeorgiev** [comment]
@danielmewes, can you send me the prebuild package to gngeorgiev.it@gmail.com, I discovered that the changefeeds break the same way in my case with rethink 2.2.5. I need at least ubuntu 14 and maybe a arch linux build. This is highly appreciated since this is huge blocker :relieved:
